### PR TITLE
[CUDA] consolidate CUDA versions

### PR DESCRIPTION
**Pull request by jameslamb**
## Changes

Proposes removing the CUDA implementation from #3160.

As of this PR, the only CUDA build of LightGBM would be the one we've been calling `cuda_exp`, which @shiyu1994 started in #4528 and #4630.

Specifically:

* removes all mentions of "CUDA exp" or "CUDA Experimental" in docs and internal code
* removes all code specific only to the implementation from #3160
* when using `python setup.py --cuda-exp` or `cmake -DUSE_CUDA_EXP=1`, raises a deprecation warning and still uses the version we've been until now calling "cuda_exp"
* removes 2 CUDA CI jobs, so now there will be three, one each for `pip`, `source`, and `wheel` builds of the CUDA-enabled Python package
* increases the minimum supported CUDA version from 9.0 to 10.0

## History

*(please correct me if I've mischaracterized the history below)*

In #3160 (*merged September 2020*), a team from IBM added a first CUDA implementation of LightGBM because the existing OpenCL-based build didn't support some platforms (namely, IBM Power).

About a year after that, @shiyu1994 and @guolinke (along with others at Microsoft?) started on an "experimental" CUDA implementation.

That "experimental" implementation was first merged in #4630 (March 2022), and since then we've had two CUDA implementations maintained in this repo:

* `cuda` = the IBM contribution
* `cuda_exp` = the newer implementation from Microsoft

Since then, @shiyu1994 has been working actively on that `cuda_exp` version, with the plan to include it in a v4.0.0 release (#5153).

The `cuda_exp` version is still missing some important features, like distributed training (#5076) and on-GPU computation of metrics and loss functions (#5163).

Despite the current limitations, this PR implements the proposal from https://github.com/microsoft/LightGBM/issues/5153#issuecomment-1369382641 to consolidate down to only one CUDA implementation in LightGBM... the one currently called `cuda_exp`.

## Motivation for this change

In my opinion, LightGBM does not have enough maintainer/contributor availability to maintain two separate CUDA implementations.

Consolidating down to 1 allows the project to more effectively channel the limited attention of its maintainers and contributors towards improving the LightGBM-on-GPU experience, by not duplicating effort across two different builds intended to serve the same purpose.

* improves development velocity by removing two costly CI jobs
* reduces confusion for users wanting to run GPU-accelerated LightGBM
* noticeably simplifies the codebase and reduces its size
* focuses all feature requests, bug reports, code contributions, etc. on one CUDA implementation

This represents a temporary loss of functionality (e.g. multi-GPU training), but I think it'll help the project to move faster and @shiyu1994 has said that that functionality is actively under development for the `cuda_exp` implementation.

## Notes for Reviewers

I know this is a large change, so tagging in others for their opinions.

@shiyu1994 @guolinke @huanzhang12 @jmoralez @StrikerRUS @btrotta @ChipKerchner @ceseo 

👋 Thanks all for your consideration.

### CODE REVIEW

**Comment by ChipKerchner** [comment]
We just want to make sure it still works (compiles, run, etc) similar to the original version.  Is there a CI building this approach?

**Comment by jameslamb** [comment]
> Is there a CI building these approaches (cuda, cuda_exp and combined)?

@ChipKerchner I don't totally understand what you mean by this question, especially "and combined". I'll try to answer but please let me know if that's not sufficient.

Every commit merged into `master` in this project in at least the last 6 months has seen the Python package built successfully and its unit tests pass for both the `cuda` version (from #3160) and `cuda_exp` version (from #4630 and onwards).

Here's the configuration for that: https://github.com/microsoft/LightGBM/blob/3c3f79e771215d25452963a8b26a4ed9c1f0589c/.github/workflows/cuda.yml#L30-L55

And build logs for the latest commit to `master`: https://github.com/microsoft/LightGBM/actions/runs/3935991924

The same CI coverage is preserved in this PR, and we'll continue to block any future PRs that break the CUDA support.

> cuda, cuda_exp, and combined

I'm confused by your use of the phrase "and combined" here, so I want to be absolutely sure you understand what's being proposed here. As of this PR, **there will only be one CUDA implementation of LightGBM**.

**Comment by jameslamb** [comment]
ah got it! Yes, the CI as of this PR tests that "one CUDA implementation of LightGBM" on Ubuntu 18.04 for the following combinations of CUDA versions, Python versions, and compilers:

https://github.com/microsoft/LightGBM/blob/967c005f7c8a2ea2cc8a2049c19c3e6a94eb68b8/.github/workflows/cuda.yml#L31-L45

you can see the build logs by clicking "Details" next to any of the checks with names starting like "CUDA Version" on this PR, e.g. https://github.com/microsoft/LightGBM/actions/runs/3945899727/jobs/6753199490

**Comment by jameslamb** [comment]
@shiyu1994 @guolinke is there any other information I could provide to help with this?

This type of PR will be difficult to keep up to date with `master` if other CUDA code is merged, since it touches so many files. I'd really like to do whatever I can to move it forward.

File: src/boosting/gbdt.h [hunk 1]
```diff
@@ -557,18 +557,18 @@ class GBDT : public GBDTBase {
   score_t* gradients_pointer_;
   /*! \brief Pointer to hessian vector, can be on CPU or GPU */
   score_t* hessians_pointer_;
-  /*! \brief Whether boosting is done on GPU, used for cuda_exp */
+  /*! \brief Whether boosting is done on GPU, used for cuda */
```
[the remark this answers is not in the document]
Comment (shiyu1994) [reply src/boosting/gbdt.h:560]: Change `cuda` to `device=cuda` for clarity.

File: src/boosting/gbdt.h [hunk 2]
```diff
@@ -557,18 +557,18 @@ class GBDT : public GBDTBase {
...
   bool boosting_on_gpu_;
-  #ifdef USE_CUDA_EXP
+  #ifdef USE_CUDA
   /*! \brief Gradient vector on GPU */
   CUDAVector<score_t> cuda_gradients_;
   /*! \brief Hessian vector on GPU */
   CUDAVector<score_t> cuda_hessians_;
-  /*! \brief Buffer for scores when boosting is on GPU but evaluation is not, used only with cuda_exp */
+  /*! \brief Buffer for scores when boosting is on GPU but evaluation is not, used only with cuda */
```
Comment (shiyu1994) [inline src/boosting/gbdt.h:567]:
```suggestion
  /*! \brief Buffer for scores when boosting is on GPU but evaluation is not, used only with device=cuda */
```

File: src/boosting/gbdt.h [hunk 3]
```diff
@@ -557,18 +557,18 @@ class GBDT : public GBDTBase {
...
   mutable std::vector<double> host_score_;
-  /*! \brief Buffer for scores when boosting is not on GPU but evaluation is, used only with cuda_exp */
+  /*! \brief Buffer for scores when boosting is not on GPU but evaluation is, used only with cuda */
```
Comment (shiyu1994) [inline src/boosting/gbdt.h:569]:
```suggestion
  /*! \brief Buffer for scores when boosting is not on GPU but evaluation is, used only with device=cuda */
```

File: src/io/config.cpp [hunk 4]
```diff
@@ -178,7 +178,12 @@ void GetDeviceType(const std::unordered_map<std::string, std::string>& params, s
     } else if (value == std::string("cuda")) {
       *device_type = "cuda";
     } else if (value == std::string("cuda_exp")) {
-      *device_type = "cuda_exp";
+      Log::Warning(
```
Comment (shiyu1994) [inline src/io/config.cpp:181]: I see. Now that we keep `cuda_exp` as an valid option with warning here, maybe it is reasonable to keep `USE_CUDA_EXP` in the make file.

File: src/treelearner/tree_learner.cpp [hunk 5]
```diff
@@ -40,24 +39,14 @@ TreeLearner* TreeLearner::CreateTreeLearner(const std::string& learner_type, con
...
-    } else if (learner_type == std::string("voting")) {
-      return new VotingParallelTreeLearner<CUDATreeLearner>(config);
-    }
-  } else if (device_type == std::string("cuda_exp")) {
     if (learner_type == std::string("serial")) {
       if (config->num_gpu == 1) {
         return new CUDASingleGPUTreeLearner(config, boosting_on_cuda);
       } else {
-        Log::Fatal("cuda_exp only supports training on a single GPU.");
+        Log::Fatal("cuda only supports training on a single GPU.");
```
Comment (shiyu1994) [inline src/treelearner/tree_learner.cpp:46]:
```suggestion
        Log::Fatal("Currently cuda version only supports training on a single GPU.");
```

File: src/treelearner/tree_learner.cpp [hunk 6]
```diff
@@ -40,24 +39,14 @@ TreeLearner* TreeLearner::CreateTreeLearner(const std::string& learner_type, con
...
       }
     } else {
-      Log::Fatal("cuda_exp only supports training on a single machine.");
+      Log::Fatal("cuda only supports training on a single machine.");
```
Comment (shiyu1994) [inline src/treelearner/tree_learner.cpp:49]:
```suggestion
      Log::Fatal("Currently cuda version only supports training on a single machine.");
```

**Review by shiyu1994 — COMMENTED** [review]
Thanks for the great work. I've viewed all the changes and just left a few comments about whether to keep `cuda_exp` as a valid input in compilation and device parameter, in addition with a few change suggestions.

File: CMakeLists.txt [hunk 7]
```diff
@@ -4,8 +4,8 @@ option(USE_GPU "Enable GPU-accelerated training" OFF)
 option(USE_SWIG "Enable SWIG to generate Java API" OFF)
 option(USE_HDFS "Enable HDFS support (EXPERIMENTAL)" OFF)
 option(USE_TIMETAG "Set to ON to output time costs" OFF)
-option(USE_CUDA "Enable CUDA-accelerated training (EXPERIMENTAL)" OFF)
-option(USE_CUDA_EXP "Enable CUDA-accelerated training with more acceleration (EXPERIMENTAL)" OFF)
+option(USE_CUDA "Enable CUDA-accelerated training " OFF)
+option(USE_CUDA_EXP "(DEPRECATED) Alias for 'USE_CUDA'. Use 'USE_CUDA' instead." OFF)
```
[the remark this answers is not in the document]
Comment (jameslamb) [reply CMakeLists.txt:8]: My initial proposal here was to keep it on `master` for a short time (maybe a few weeks) so that people cloning the repo without pinning to a specific commit would at least get an informative warning. But revisiting it I think you're right...we should just remove `CUDA_EXP` completely everywhere. I'll push another commit here doing that.

File: src/boosting/gbdt.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (jameslamb) [reply src/boosting/gbdt.h:560]: sure! Although since `device` is just an alias for main parameter `device_type` (https://lightgbm.readthedocs.io/en/latest/Parameters.html#device_type), I think we should use `device_type`.

```suggestion
  /*! \brief Whether boosting is done on GPU, used for device_type=cuda */
```

File: src/boosting/gbdt.h (hunk 2 above)
Comment (jameslamb) [reply src/boosting/gbdt.h:567]:
```suggestion
  /*! \brief Buffer for scores when boosting is on GPU but evaluation is not, used only with device_type=cuda */
```

File: src/boosting/gbdt.h (hunk 3 above)
Comment (jameslamb) [reply src/boosting/gbdt.h:569]:
```suggestion
  /*! \brief Buffer for scores when boosting is not on GPU but evaluation is, used only with device_type=cuda */
```

File: docs/Parameters.rst [hunk 8]
```diff
@@ -205,7 +205,7 @@ Core Parameters
 
    -  **Note**: please **don't** change this during training, especially when running multiple jobs simultaneously by external packages, otherwise it may cause undesirable errors
 
--  ``device_type`` :raw-html:`<a id="device_type" title="Permalink to this parameter" href="#device_type">&#x1F517;&#xFE0E;</a>`, default = ``cpu``, type = enum, options: ``cpu``, ``gpu``, ``cuda``, ``cuda_exp``, aliases: ``device``
+-  ``device_type`` :raw-html:`<a id="device_type" title="Permalink to this parameter" href="#device_type">&#x1F517;&#xFE0E;</a>`, default = ``cpu``, type = enum, options: ``cpu``, ``gpu``, ``cuda``, aliases: ``device``
 
    -  device for the tree learning, you can use GPU to achieve the faster learning
```
[the remark this answers is not in the document]
Comment (jameslamb) [reply docs/Parameters.rst]: Good idea! I included an expansion of this documentation in https://github.com/microsoft/LightGBM/pull/5677/commits/ac7ab774d5152478aa660028c2ee0d866a008daa, that I think will help users to choose between these options.

File: CMakeLists.txt (hunk 7 above)
[the remark this answers is not in the document]
Comment (jameslamb) [reply CMakeLists.txt]: I just pushed https://github.com/microsoft/LightGBM/pull/5677/commits/ac7ab774d5152478aa660028c2ee0d866a008daa completely removing remaining uses of `CUDA_EXP`.

File: src/io/config.cpp (hunk 4 above)
Comment (jameslamb) [reply src/io/config.cpp]: I just pushed https://github.com/microsoft/LightGBM/commit/ac7ab774d5152478aa660028c2ee0d866a008daa completely removing remaining uses of `CUDA_EXP`, based on https://github.com/microsoft/LightGBM/pull/5677#discussion_r1089889981

File: python-package/setup.py [hunk 9]
```diff
@@ -21,7 +21,7 @@
     ('integrated-opencl', None, 'Compile integrated OpenCL version'),
     ('gpu', 'g', 'Compile GPU version'),
     ('cuda', None, 'Compile CUDA version'),
-    ('cuda-exp', None, 'Compile CUDA Experimental version'),
+    ('cuda-exp', None, '(deprecated) Alias for "cuda". Use "cuda" instead.'),
```
[the remark this answers is not in the document]
Comment (jameslamb) [reply python-package/setup.py]: I just pushed https://github.com/microsoft/LightGBM/commit/ac7ab774d5152478aa660028c2ee0d866a008daa completely removing remaining uses of `CUDA_EXP`, based on https://github.com/microsoft/LightGBM/pull/5677#discussion_r1089889981

File: CMakeLists.txt [hunk 10]
```diff
@@ -31,7 +31,7 @@ elseif(USE_SWIG)
   cmake_minimum_required(VERSION 3.8)
 elseif(USE_GPU OR APPLE)
   cmake_minimum_required(VERSION 3.2)
-elseif(USE_CUDA OR USE_CUDA_EXP)
+elseif(USE_CUDA)
```
[the remark this answers is not in the document]
Comment (jameslamb) [reply CMakeLists.txt:33]: I just pushed https://github.com/microsoft/LightGBM/commit/ac7ab774d5152478aa660028c2ee0d866a008daa completely removing remaining uses of `CUDA_EXP`, based on https://github.com/microsoft/LightGBM/pull/5677#discussion_r1089889981

**Comment by jameslamb** [comment]
Thanks @shiyu1994 . I just pushed some commits completely removing `CUDA_EXP`, instead of keeping that configuration option and raising a warning. I think that'll reduce confusion, and it's ok given that it was never officially included in a release.

**Comment by shiyu1994** [comment]
@jameslamb Thanks. The changes LGTM. It seems that we are encountering some CI issues. One R test failes. And one gpu job in Azure Devops ci test fails. I've tried retrigger the R test again but it fails again. Maybe we should fix the ci issue first.

**Comment by jameslamb** [comment]
Thanks!

@shiyu1994 I've already put up a PR to fix the R CI issues. Can you please review #5689?

I've noticed that the GPU jobs on Azure DevOps have gotten flakier since I merged #5292. It's usually fixed by re-running once or twice. I'll keep doing that. We can turn the Dask tests back off on GPU builds in the future if it gets too annoying or we don't have time to investigate the issues.

**Comment by shiyu1994** [comment]
> I've already put up a PR to fix the R CI issues. Can you please review https://github.com/microsoft/LightGBM/pull/5689?

Sorry for the delay. I see that PR is already merged. Thanks.

> I've noticed that the GPU jobs on Azure DevOps have gotten flakier since I merged https://github.com/microsoft/LightGBM/pull/5292.

I agree. But I'll try to spare some time to investigate it if it happens frequently.

**Comment by jameslamb** [comment]
no problem, thanks @shiyu1994 ! I just merged in the changes from #5689 (thanks to @jmoralez for reviewing!) here so once CI rebuilds on this PR, and if you approve this PR, I think we can merge it.

**Review by shiyu1994 — APPROVED** [review]
Thanks. The changes LGTM.

**Comment by jameslamb** [comment]
awesome, thank you so much @shiyu1994 !!! I'm excited to have shorter CI times and for us to be able to focus on a single CUDA version 😁 

And thank you so much @ChipKerchner and your teammates for getting LightGBM started on this CUDA journey back in #3160.
